### DISCUSSION TITLE
Suggestion: Replacing with IWindow as new name IGameWindow? And new ISystemWindow will release soon.

### DISCUSSION THREAD

**Discussion by DeafMan1983** [dotnet/Silk.NET#531] [Ideas] [answered]
Hello friends,

I have idea #IWindow should replace with new name #IGameWindow 
And I made ISystemWindow, it is OS-engine-able GUI like WinForms or Cocoa and X11 / Wayland and Android's UI etc..

Example:
I create ISystemWindow embedding in IGameWindow.
```
    class Program
    {
        private static ISystemWindow _swindow;
        private static IGameWindow _gwindow; // Replaced from IWindow

        ...


        static void Main(string[] args)
        {
            var options = new WindowOptions();
            ...
            _gwindow = GameWindow.Create(options); // Replaced from Window
           _swindow = SystemWindow.Create(_gwindow, 250, _gwindow.Size.Y);
            ...
            _swindow.Map();
           _gwindow.Run();
        }

        ...

        public static OnClose()
        {
            _swindow.Destroy();
            _gwindow.Close();
        }

        ...
    }
```
ISystemWindow will work all operating system via RuntimeInformation like you use Windows 10 for example.

Thank you and I hope you think my idea is great or not good?

**Comment by HurricanKai** [comment]
I don't really understand what the advantages are here, please provide more details.

**Reply by DeafMan1983** [reply]
Entschuldige mich bitte @HurricanKai !
I explain like multiple window with GUI ( SystemWindow ) and GameWindow or embedding GUI ( SystemWindow ) inside GameWindow 

If I am finish then I will show you. And you will experience that.

**Comment by Perksey** [comment, marked as answer]
Struggling to understand here as well, if you'd like to propose something please gather as many details as you can and then use the proposals system as outlined here: https://github.com/dotnet/Silk.NET/tree/main/documentation/proposals

You can see the original windowing proposal here, though it's a bit out of date now: https://github.com/dotnet/Silk.NET/blob/main/documentation/proposals/Proposal%20-%20Windowing.md

One thing we can't do, however, is this embedding idea you propose as we can't do this on both GLFW and SDL. We recommend users either use viewport trickery or some other window system to achieve something like this.

If it passes preliminary review by Kai and myself we'll pass it along to the Working Group for discussion in a Working Group Meeting.

**Reply by DeafMan1983** [reply]
Hello friends, sorry I am very busy with X11, GLX and VK for SystemWindow .
![image](https://user-images.githubusercontent.com/57066679/123522343-f656a000-d6bc-11eb-9e88-99bca3625662.png)

PS: I enjoy and take rest in weekend. 

I have implementation for MM and inch switcher 
Example:
```
            // with "//" then Width and Height will in inch format
            display.EnableMillimeters(); // If you enable for mm format
            Console.WriteLine("Width in mm: " + display.Width);
            Console.WriteLine("Height in mm: " + display.Height);

            Screen screen = display.Screen;
            
            // with "//" then Width and Height will in inch format
            // screen.EnableMillimeters(); // If you enable for mm format
            Console.WriteLine("Width: " + screen.Width);
            Console.WriteLine("Height: " + screen.Height);
```
It looks result:
```
Width in mm: 508
Height in mm: 285
Width: 1920
Height: 1080
```
in Silk.NET.X11/Display.cs and Silk.NET.X11/Screen.cs I have added method for enabling millimeters if you don't like millimeters then you should ignore to add method. The measure will shows normal inch.

Enjoy weekend!

**Reply by DeafMan1983** [reply]
Hello my friends sorry for longer time. 

I am developing one [LibraryLoader ](https://github.com/DeafMan1983/LibraryLoader-for-Net-6.0) ( It is alternative NativeApi ). I want tell you why do I need add Windows64 and Windows86 for Windows in SearchPathContainer.cs - If you generate end-of-user dotnot public package then public package can work current arch like arm, x64 or x86 See [RID of Microsoft's website](https://docs.microsoft.com/de-de/dotnet/core/rid-catalog) I don't need set x64 or x86.

Now bad news for WinAPI because WinApi cannot dock child in SDL2 - So sadly - See [SteamPowered Community](https://steamcommunity.com/discussions/forum/24/3044984779779043883/) I don't expect that Half-Life has GUI like dialog, Button etc inside SDL2 - Do you have chance to WinApi can dock inside SDL2- or GLFW3-Window? Like SetParent() and  SetWindowPos() see [flibitijibibo's hooking SDL2](https://gist.github.com/flibitijibibo/cf282bfccc1eaeb47550) inside WinForms but it looks wrong - I need choose "WinForm or WinApi inside SDL2"
I don't know how did Valve Software put GUI ( called VGUI ) inside SDL2 for Half-Life Launcher.
For X11 and Wayland have not problem to dock child-inside-SDL2-Window. But why does X11/Wayland make happy then WinApi (User32)?

// EDIT:
Is it possible for advanced custom SDL_Window with WinApi 8 built-in MDI-Child inside SDL_Window like SDL_CreateWindowFrom() ? But it takes giant work.

In Half-Life's SDK - I can't find SDL_GetWindowWMInfo() - Is it like IMGUI or what?

Kind Regards!

**Reply by Perksey** [reply]
I have no idea how Half-Life's GUI works if that's the question? Silk.NET has no support for child windows today and there are currently no plans to introduce support for them due to cross-platform concerns.

**Reply by DeafMan1983** [reply]
You're right. I mean GUI of Half-Life, NanoGUI or IMGUI for Silk.NET. That is why I look for advanced way. Thank you for support! And I will create own.
Best wishes!
